### DISCUSSION TITLE
"my TYPE VARLIST" isn't good enough

### DISCUSSION THREAD

**Discussion by Ovid** [Perl-Oshun/oshun#40] [Q&A]
From `perldoc -f my`:

```
    my VARLIST
    my TYPE VARLIST
    my VARLIST : ATTRS
    my TYPE VARLIST : ATTRS
```

Currently, we can write code like this:

```perl
my ( $count, $aref ) = @_;
```

The `TYPE` cannot handle that. It assumes one type for all.

If a `VAR` is defined as `VAR` or `TYPE VAR`, we can make this work:

```perl
my ( UINT $count, ARRAY $aref ) = @_;
```

(Or using the attribute syntax, perhaps).

Paging @Perl-Oshun/core-team and @leonerd 

Thoughts?

**Comment by druud** [comment]
I presume this compiles:
```
(my UINT $count, my ARRAY $aref) = @_;
```
but I always wondered how the first my doesn't see a VARLIST of 2 elements.

This certainly doesn't compile:
```
my main ($i, my CORE @r) = @_; 
```
(ignore the main/CORE usage, they are only there as available classnames)

**Reply by Ovid** [reply]
Don't know why it doesn't see a `VARLIST`, but I think people would strongly object to that being the syntax to use for checks.

**Comment by perigrin** [comment]
(Using the word type here cause that’s how perldocs refer to them)

I’m comfortable with changing the type annotations the way you’ve suggested … this is effectively how it’s going to have to change to allow type annotations in signatures. 

If `sub ($arg, @rest) {…}` is supposed to be equivalent to `sub { my ($arg, @rest) = @_; … }` then we are gonna need to alter the way type annotations are parsed if we want type annotations in signatures.

I think `sub { my ( UINT $count, ARRAY $aref ) = @_; }` is a natural expansion of `sub ( UINT $count, ARRAY $aref ) {}` … so logic dictates if the latter is allowed so is the former.

**Reply by zmughal** [reply]
Also would align with [destructuring assignment](https://github.com/orgs/Perl-Oshun/discussions/22#discussioncomment-6521058).

**Reply by perigrin** [reply]
Exactly.
